File: sdks/ts/memory/src/ingest/schedule/cron.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Minimal cron expression parser supporting standard 5-field syntax:
 * minute, hour, day-of-month, month, day-of-week.
 *
 * Supports: numbers, ranges (1-5), steps (* /5), lists (1,3,5), and *.
 */

export type CronSchedule = {
  readonly minute: readonly number[]
  readonly hour: readonly number[]
  readonly dayOfMonth: readonly number[]
  readonly month: readonly number[]
  readonly dayOfWeek: readonly number[]
  /** True when the original day-of-month token was '*' (wildcard). */
  readonly dayOfMonthIsWild: boolean
  /** True when the original day-of-week token was '*' (wildcard). */
  readonly dayOfWeekIsWild: boolean
}

export const parseCron = (expression: string): CronSchedule => {
  const fields = expression.trim().split(/\s+/)
  if (fields.length !== 5) {
    throw new Error(`cron: expected 5 fields, got ${fields.length} in "${expression}"`)
  }

  const [minuteField, hourField, domField, monthField, dowField] = fields as [string, string, string, string, string]

  return {
    minute: parseField(minuteField, 0, 59, 'minute'),
    hour: parseField(hourField, 0, 23, 'hour'),
    dayOfMonth: parseField(domField, 1, 31, 'day-of-month'),
    month: parseField(monthField, 1, 12, 'month'),
    dayOfWeek: parseField(dowField, 0, 6, 'day-of-week'),
    dayOfMonthIsWild: isWildcard(domField),
    dayOfWeekIsWild: isWildcard(dowField),
  }
}

/**
 * Returns true when the raw cron field token represents a wildcard.
 * Matches bare '*' and step-only forms like '* /N' (without space).
 */
const isWildcard = (token: string): boolean => {
  const trimmed = token.trim()
  return trimmed === '*' || trimmed.startsWith('*/')
}

export const isValid = (expression: string): boolean => {
  try {
    parseCron(expression)
    return true
  } catch {
    return false
  }
}

/**
 * Compute the next occurrence of the schedule after the given reference
 * time. Searches up to 4 years ahead.
 *
 * DOM+DOW union semantics: per POSIX cron, when both day-of-month and
 * day-of-week are non-wildcard (not full-range), a date matches if
 * EITHER condition is true. When one or both are wildcard, standard
 * intersection (AND) applies.
 */
export const nextOccurrence = (sched: CronSchedule, after: Date): Date => {
  // Start one minute after `after`, zeroing seconds.
  const start = new Date(after)
  start.setSeconds(0, 0)
  start.setMinutes(start.getMinutes() + 1)

  const minuteSet = toSet(sched.minute)
  const hourSet = toSet(sched.hour)
  const domSet = toSet(sched.dayOfMonth)
  const monthSet = toSet(sched.month)
  const dowSet = toSet(sched.dayOfWeek)

  const domIsWild = sched.dayOfMonthIsWild
  const dowIsWild = sched.dayOfWeekIsWild

  const limit = new Date(after)
  limit.setFullYear(limit.getFullYear() + 4)

  const t = new Date(start)
  while (t < limit) {
    const mon = t.getMonth() + 1 // JS months are 0-indexed
    if (!monthSet.has(mon)) {
      t.setMonth(t.getMonth() + 1)
      t.setDate(1)
      t.setHours(0, 0, 0, 0)
      continue
    }
    const dayMatch = matchDay(domSet, dowSet, domIsWild, dowIsWild, t.getDate(), t.getDay())
    if (!dayMatch) {
      t.setDate(t.getDate() + 1)
      t.setHours(0, 0, 0, 0)
      continue
    }
    if (!hourSet.has(t.getHours())) {
      t.setHours(t.getHours() + 1)
      t.setMinutes(0, 0, 0)
      continue
    }
    if (!minuteSet.has(t.getMinutes())) {
      t.setMinutes(t.getMinutes() + 1)
      t.setSeconds(0, 0)
      continue
    }
    return new Date(t)
  }

  return limit
}

/**
 * POSIX cron union semantics for day matching. When both DOM and DOW
 * are restricted (non-wildcard), match if EITHER is true. Otherwise
 * use standard AND logic.
 */
const matchDay = (
  domSet: Set<number>,
  dowSet: Set<number>,
  domIsWild: boolean,
  dowIsWild: boolean,
  day: number,
  weekday: number,
): boolean => {
  if (!domIsWild && !dowIsWild) {
    return domSet.has(day) || dowSet.has(weekday)
  }
  return domSet.has(day) && dowSet.has(weekday)
}

type CronRange = { readonly start: number; readonly end: number }

/**
 * Parses a single range token from a cron field. Handles three forms:
 * wildcard ('*'), explicit range ('1-5'), or single value ('3').
 */
const parseCronRange = (rangePart: string, min: number, max: number, name: string): CronRange => {
  if (rangePart === '*') {
    return { start: min, end: max }
  }

  if (rangePart.includes('-')) {
    const rangeSplit = rangePart.split('-')
    const s = rangeSplit[0] ?? ''
    const e = rangeSplit[1] ?? ''
    const start = Number.parseInt(s, 10)
    const end = Number.parseInt(e, 10)
    if (Number.isNaN(start) || Number.isNaN(end)) {
      throw new Error(`cron: ${name} field: invalid range "${rangePart}"`)
    }
    return { start, end }
  }

  const val = Number.parseInt(rangePart, 10)
  if (Number.isNaN(val)) {
    throw new Error(`cron: ${name} field: invalid value "${rangePart}"`)
  }
  return { start: val, end: val }
}

const parseField = (field: string, min: number, max: number, name: string): number[] => {
  const result: number[] = []

  for (const part of field.split(',')) {
    const trimmed = part.trim()
    const stepParts = trimmed.split('/')
    const rangePart = stepParts[0] ?? ''
    let step = 1

    if (stepParts.length === 2) {
      const stepStr = stepParts[1] ?? ''
      step = Number.parseInt(stepStr, 10)
      if (Number.isNaN(step) || step < 1) {
        throw new Error(`cron: ${name} field: invalid step "${stepStr}"`)
      }
    }

    const { start: rangeStart, end: rangeEnd } = parseCronRange(rangePart, min, max, name)

    if (rangeStart < min || rangeEnd > max || rangeStart > rangeEnd) {
      throw new Error(`cron: ${name} field: value out of range [${min}-${max}]: ${rangeStart}-${rangeEnd}`)
    }

    for (let i = rangeStart; i <= rangeEnd; i += step) {
      result.push(i)
    }
  }

  if (result.length === 0) {
    throw new Error(`cron: ${name} field: empty`)
  }

  return [...new Set(result)]
}

const toSet = (values: readonly number[]): Set<number> => new Set(values)

File: go/ingest/schedule/cron.go
// SPDX-License-Identifier: Apache-2.0
package schedule

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

// CronSchedule represents a parsed 5-field cron expression.
type CronSchedule struct {
	Minute     []int // 0-59
	Hour       []int // 0-23
	DayOfMonth []int // 1-31
	Month      []int // 1-12
	DayOfWeek  []int // 0-6 (Sunday=0)
}

// ParseCron parses a standard 5-field cron expression.
// Supports: numbers, ranges (1-5), steps (*/5), lists (1,3,5), and *.
func ParseCron(expression string) (CronSchedule, error) {
	fields := strings.Fields(expression)
	if len(fields) != 5 {
		return CronSchedule{}, fmt.Errorf("cron: expected 5 fields, got %d in %q", len(fields), expression)
	}

	minute, err := parseField(fields[0], 0, 59)
	if err != nil {
		return CronSchedule{}, fmt.Errorf("cron: minute field: %w", err)
	}
	hour, err := parseField(fields[1], 0, 23)
	if err != nil {
		return CronSchedule{}, fmt.Errorf("cron: hour field: %w", err)
	}
	dom, err := parseField(fields[2], 1, 31)
	if err != nil {
		return CronSchedule{}, fmt.Errorf("cron: day-of-month field: %w", err)
	}
	month, err := parseField(fields[3], 1, 12)
	if err != nil {
		return CronSchedule{}, fmt.Errorf("cron: month field: %w", err)
	}
	dow, err := parseField(fields[4], 0, 6)
	if err != nil {
		return CronSchedule{}, fmt.Errorf("cron: day-of-week field: %w", err)
	}

	return CronSchedule{
		Minute:     minute,
		Hour:       hour,
		DayOfMonth: dom,
		Month:      month,
		DayOfWeek:  dow,
	}, nil
}

// IsValid reports whether expression is a valid 5-field cron expression.
func IsValid(expression string) bool {
	_, err := ParseCron(expression)
	return err == nil
}

// NextOccurrence computes the next time the schedule fires after the
// given reference time. It iterates minute-by-minute up to 4 years out
// to find the next match.
//
// DOM+DOW union semantics: per POSIX cron, when both day-of-month and
// day-of-week are non-wildcard (i.e. not full-range), a date matches if
// EITHER condition is true. When one or both are wildcard, standard
// intersection applies.
func NextOccurrence(sched CronSchedule, after time.Time) time.Time {
	// Start one minute after `after`, zeroing seconds.
	t := after.Truncate(time.Minute).Add(time.Minute)

	minuteSet := toSet(sched.Minute)
	hourSet := toSet(sched.Hour)
	domSet := toSet(sched.DayOfMonth)
	monthSet := toSet(sched.Month)
	dowSet := toSet(sched.DayOfWeek)

	domIsWild := len(sched.DayOfMonth) == 31
	dowIsWild := len(sched.DayOfWeek) == 7

	// Search up to 4 years (≈2.1M minutes) to find a match.
	limit := after.Add(4 * 365 * 24 * time.Hour)
	for t.Before(limit) {
		dayMatch := matchDay(domSet, dowSet, domIsWild, dowIsWild, t.Day(), int(t.Weekday()))
		if monthSet[int(t.Month())] &&
			dayMatch &&
			hourSet[t.Hour()] &&
			minuteSet[t.Minute()] {
			return t
		}

		// Skip ahead intelligently.
		if !monthSet[int(t.Month())] {
			// Jump to the first day of the next month.
			t = time.Date(t.Year(), t.Month()+1, 1, 0, 0, 0, 0, t.Location())
			continue
		}
		if !dayMatch {
			// Jump to the next day.
			t = time.Date(t.Year(), t.Month(), t.Day()+1, 0, 0, 0, 0, t.Location())
			continue
		}
		if !hourSet[t.Hour()] {
			// Jump to the next hour.
			t = time.Date(t.Year(), t.Month(), t.Day(), t.Hour()+1, 0, 0, 0, t.Location())
			continue
		}
		t = t.Add(time.Minute)
	}

	// Fallback: should not happen for valid schedules.
	return limit
}

// matchDay implements POSIX cron union semantics for day-of-month and
// day-of-week. When both fields are restricted (non-wildcard), the day
// matches if EITHER the DOM or DOW matches. When one or both are
// wildcard, standard AND logic applies.
func matchDay(domSet, dowSet map[int]bool, domIsWild, dowIsWild bool, day, weekday int) bool {
	if !domIsWild && !dowIsWild {
		return domSet[day] || dowSet[weekday]
	}
	return domSet[day] && dowSet[weekday]
}

// parseField parses a single cron field (e.g., "*/5", "1-3", "1,3,5", "*").
func parseField(field string, min, max int) ([]int, error) {
	var result []int

	for _, part := range strings.Split(field, ",") {
		part = strings.TrimSpace(part)
		stepParts := strings.SplitN(part, "/", 2)
		rangePart := stepParts[0]
		step := 1

		if len(stepParts) == 2 {
			var err error
			step, err = strconv.Atoi(stepParts[1])
			if err != nil || step < 1 {
				return nil, fmt.Errorf("invalid step %q", stepParts[1])
			}
		}

		var rangeStart, rangeEnd int

		if rangePart == "*" {
			rangeStart = min
			rangeEnd = max
		} else if idx := strings.Index(rangePart, "-"); idx >= 0 {
			var err error
			rangeStart, err = strconv.Atoi(rangePart[:idx])
			if err != nil {
				return nil, fmt.Errorf("invalid range start %q", rangePart[:idx])
			}
			rangeEnd, err = strconv.Atoi(rangePart[idx+1:])
			if err != nil {
				return nil, fmt.Errorf("invalid range end %q", rangePart[idx+1:])
			}
		} else {
			val, err := strconv.Atoi(rangePart)
			if err != nil {
				return nil, fmt.Errorf("invalid value %q", rangePart)
			}
			rangeStart = val
			rangeEnd = val
		}

		if rangeStart < min || rangeEnd > max || rangeStart > rangeEnd {
			return nil, fmt.Errorf("value out of range [%d-%d]: %d-%d", min, max, rangeStart, rangeEnd)
		}

		for i := rangeStart; i <= rangeEnd; i += step {
			result = append(result, i)
		}
	}

	if len(result) == 0 {
		return nil, fmt.Errorf("empty field")
	}
	return deduplicate(result), nil
}

// deduplicate removes duplicate values from a sorted-ish slice
// produced by parseField, preserving order.
func deduplicate(vals []int) []int {
	seen := make(map[int]bool, len(vals))
	out := make([]int, 0, len(vals))
	for _, v := range vals {
		if !seen[v] {
			seen[v] = true
			out = append(out, v)
		}
	}
	return out
}

func toSet(vals []int) map[int]bool {
	s := make(map[int]bool, len(vals))
	for _, v := range vals {
		s[v] = true
	}
	return s
}
